src/inatinqperf/bench/bench.py

```diff
@@ -0,0 +1,319 @@
+    """Ensure directory exists."""
+    Path(p).mkdir(parents=True, exist_ok=True)
+    return p
+
+
+def exact_baseline(x: np.ndarray, metric: str) -> faiss.Index:
+    """Exact baseline index using FAISS IndexFlat*."""
+    d = x.shape[1]
+    base = faiss.IndexFlatIP(d) if metric in ("ip", "cosine") else faiss.IndexFlatL2(d)
+    index = faiss.IndexIDMap2(base)
+    ids = np.arange(x.shape[0], dtype="int64")
+    index.add_with_ids(x.astype("float32"), ids)
+    return index
+
+
+def recall_at_k(approx_i: np.ndarray, exact_i: np.ndarray, k: int) -> float:
+    """Compute recall@K between two sets of indices."""
+    hits = 0
+    for i in range(approx_i.shape[0]):
+        hits += len(set(approx_i[i, :k]).intersection(set(exact_i[i, :k])))
+    return hits / float(approx_i.shape[0] * k)
+
+
+def cmd_download(args: argparse.Namespace, cfg: Mapping[str, Any]) -> None:
+    """Download HF dataset and optionally export images."""
+    hf_id = cfg["dataset"]["hf_id"]
+    size = args.size
+    out_dir = args.out_dir or cfg["dataset"]["out_dir"]
+    export = (
+        args.export_images if args.export_images is not None else cfg["dataset"].get("export_images", True)
+    )
+    split_map = cfg["dataset"]["size_splits"]
+    split = split_map.get(size, split_map["small"])
+    ensure_dir(out_dir)
+    with Profiler(f"download-{hf_id}-{size}"):
+        ds = load_composite(hf_id, split)
+        ds.save_to_disk(out_dir)
+        if export:
+            export_dir = Path(out_dir) / "images"
+            manifest = export_images(ds, export_dir)
+            print(f"Exported images to: {export_dir}\nManifest: {manifest}")
+    print(f"Saved HF dataset to: {out_dir}")
+
+
+def cmd_embed(args: argparse.Namespace, cfg: Mapping[str, Any]) -> None:
+    """Compute CLIP embeddings and save HF dataset with 'embedding'."""
+    model_id = args.model_id or cfg["embedding"]["model_id"]
+    batch = args.batch or int(cfg["embedding"]["batch"])
+    raw_dir = args.raw_dir or cfg["dataset"]["out_dir"]
+    emb_dir = args.emb_dir or cfg["embedding"]["out_dir"]
+    out_hf_dir = cfg["embedding"]["out_hf_dir"]
+    ensure_dir(emb_dir)
+    ensure_dir(out_hf_dir)
+    with Profiler("embed-images"):
+        _, x, ids, labels = embed_images(raw_dir, model_id, batch)
+        ds2 = to_hf_dataset(x, ids, labels)
+        ds2.save_to_disk(out_hf_dir)
+        print(f"Embeddings: {x.shape} -> {out_hf_dir}")
+
+
+def _init_backend(backend_name: str, dim: int, metric: str, params: dict[str, Any]) -> dict:
+    """Instantiate backend and call init(), scrubbing reserved keys from params.
+
+    Prevents errors like: TypeError: ... init() got multiple values for keyword 'metric'.
+    """
+    cls = ALL_BACKENDS[backend_name]
+    be = cls()
+    # Avoid passing duplicate values for explicit kwargs
+    safe_params = dict(params) if params else {}
+    for k in ("metric", "dim", "name"):
+        safe_params.pop(k, None)
+    be.init(dim=dim, metric=metric, **safe_params)
```

Comment:
Don't require two-phase construction. Pass `**safe_params` to the `cls`'s constructor.